src/platform_impl/windows/event_loop.rs

```diff
@@ -393,57 +425,84 @@ impl EventLoop {
         //
         runner.prepare_wait();
 
-        let control_flow_timeout = match runner.control_flow() {
-            ControlFlow::Wait => None,
-            ControlFlow::Poll => Some(Duration::ZERO),
-            ControlFlow::WaitUntil(wait_deadline) => {
-                let start = Instant::now();
-                Some(wait_deadline.saturating_duration_since(start))
-            },
-        };
-        let timeout = min_timeout(control_flow_timeout, timeout);
+        {
+            let control_flow_timeout = match runner.control_flow() {
+                ControlFlow::Wait => None,
+                ControlFlow::Poll => Some(Duration::ZERO),
+                ControlFlow::WaitUntil(wait_deadline) => {
+                    let start = Instant::now();
+                    Some(wait_deadline.saturating_duration_since(start))
+                },
+            };
+            timeout = min_timeout(timeout, control_flow_timeout);
+        }
 
-        // # Safety
-        // The Windows API has no documented requirement for bitwise
-        // initializing a `MSG` struct (it can be uninitialized memory for the C
-        // API) and there's no API to construct or initialize a `MSG`. This
-        // is the simplest way avoid uninitialized memory in Rust
-        let mut msg = unsafe { mem::zeroed() };
-        let msg_status = wait_for_msg(&mut msg, timeout);
```

Comment:
Why did we stop using `wait_for_msg`?

Reply:
Because I rewrote code to use `MsgWaitForMultipleObjectsEx` instead of `GetMessageW`. `MsgWaitForMultipleObjectsEx` only wait until a new message is available but doesn't take it and doesn't remove it from queue, while `GetMessageW` does wait and take it.

In new version, `wait_for_messages` only waits until message is available, then `dispatch_peeked_messages` takes them from queue and dispatches, until the queue is empty.

So, as I understand, old version of code needed to manage both waiting and dispatching messages so it was split it into subfunctions. New version of code only does waiting while all dispatching done by `dispatch_peeked_messages` so there is no need to create a subfunction which just repeats the name of outer function `wait_for_messages`.

Also, after I move inner functions out of `wait_for_messages`, as you requested in https://github.com/rust-windowing/winit/pull/3950#discussion_r1798559529 , `wait_for_messages` would be tiny and splitting it further would be pointless.

Reply:
I believe there are some messages that `MsgWaitForMultipleObjectsEx` does not receive, which is why we used the timer system and `GetMessageW` here. 

cc @rib you were the one to set up this system

Reply:
I don't really recall there being a particular issue with `MsgWaitForMultipleObjectsEx` and at the time I suppose it just seemed simpler to use `GetMessage` while we were only waiting for messages and not waiting on any object handles.

The main concern I had when last iterating on this code was to do with the indirection of a dedicated waiter thread and there were some funky details to do with redraw handling I don't really recall.

Ref: fae4cbd2aa67315b29f139142bb54a5ff358b95c and https://github.com/rust-windowing/winit/pull/2767

This change to use hrtimers seems good to me, and can see that it makes sense to use `MsgWaitForMultipleObjectsEx` to integrate the timer handle.

It seems to make sense that the waiting is now decoupled, since `MsgWaitForMultipleObjectsEx` doesn't take any data when it has something to wake up for - unlike `GetMessageW` which would wait _and_ take.